src/devices/bus/coco/coco_fdc.cpp

```diff
@@ -694,7 +466,323 @@ DEFINE_DEVICE_TYPE_PRIVATE(COCO_FDC_V11, coco_family_fdc_device_base, coco_fdc_v
+	{
+		set_line_value(line::HALT, CLEAR_LINE);
+
+		if (drq() == ASSERT_LINE)
+		{
+			if ((m_cache_controler & 0x01) == 0x01) /* Read cache on */
+			{
+				u8 data = m_wd17xx->data_r();
+				LOGWDSCII("cache drq read: %02x\n", data );
+				m_cache[m_cache_pointer++] = data;
+			}
+			else /* Write cache on */
+			{
+				u8 data = m_cache[m_cache_pointer++];
+				LOGWDSCII("cache drq write: %02x\n", data );
+				m_wd17xx->data_w(data);
+			}
+
+			m_cache_pointer &= 0x1ff;
+		}
+	}
+
+	if ((m_cache_controler & 0x08) == 0x08)
+	{
+		m_carts->in_w<0>(intrq());
+	}
+	else
+	{
+		m_carts->in_w<0>(CLEAR_LINE);
+	}
+
+	if ((m_cache_controler & 0x04) == 0x00)
+	{
+		set_line_value(line::NMI, intrq() && (dskreg() & 0x20));
+	}
+	else
+	{
+		set_line_value(line::NMI, CLEAR_LINE);
+	}
+}
+
+//-------------------------------------------------
+//	ff74_read - no halt registers
+//-------------------------------------------------
+
+u8 coco_scii_device_base::ff74_read(offs_t offset)
+{
+	u8 data = 0x0;
+
+	switch(offset)
+	{
+		case 0x0:
+		case 0x1:
+			data = m_cache[m_cache_pointer++];
+			LOGWDSCII("cache read: %04x = %02x\n", m_cache_pointer, data);
+			m_cache_pointer &= 0x1ff;
+			break;
+
+		case 0x2:
+		case 0x3:
+			if (intrq() == ASSERT_LINE)
+			{
+				m_cache_controler &= 0x7f;
+			}
+			else
+			{
+				m_cache_controler |= 0x80;
+			}
+
+			data = m_cache_controler;
+			LOGWDSCII("control read:  %02x\n", data);
+			break;
+	}
+
+	return data;
+}
+
+//-------------------------------------------------
+//	ff74_write - no halt registers
+//-------------------------------------------------
+
+void coco_scii_device_base::ff74_write(offs_t offset, u8 data)
+{
+	switch(offset)
+	{
+		case 0x0:
+		case 0x1:
+			LOGWDSCII("cache write: %04x = %02x\n", m_cache_pointer, data);
+			m_cache[m_cache_pointer++] = data;
+			m_cache_pointer &= 0x1ff;
+			break;
+		case 0x2:
+		case 0x3:
+			LOGWDSCII("control write: %02x\n", data);
+
+			// reset static ram buffer pointer on any write
+			m_cache_pointer = 0;
+
+			m_cache_controler = data;
+			update_lines();
+			break;
+	}
+
+}
+
+
+//**************************************************************************
+//	Disto / CRC  Super Controller II, CoCo 1 / 2 ROM
+//**************************************************************************
+
+ROM_START(coco_scii_cc1)
+	ROM_REGION(0x8000, "eprom", ROMREGION_ERASE00)
+	ROM_LOAD("cdos 4 4-6-89 cc1.bin", 0x0000, 0x4000, CRC(9da6db28) SHA1(2cc6e275178ca8d8f281d845792fb0ae069aaeda))
+ROM_END
+
+namespace
+{
+	class coco_scii_device_cc1 : public coco_scii_device_base
+	{
+	public:
+		// construction/destruction
+		coco_scii_device_cc1(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+			: coco_scii_device_base(mconfig, COCO_SCII_CC1, tag, owner, clock)
+		{
+		}
+
+	protected:
+		// optional information overrides
+		virtual const tiny_rom_entry *device_rom_region() const override
+		{
+			return ROM_NAME(coco_scii_cc1);
+		}
+
+	};
+
+}
```

Comment:
Rather than doing this, please just put all the local classes and their implementations inside an anonymous namespace.  It’s far less error-prone and easier for a human to understand.

Reply:
Fixed. I think.